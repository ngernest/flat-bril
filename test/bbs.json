{
  "functions": [
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        },
        {
          "name": "modulus",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "n",
            "modulus"
          ],
          "dest": "result",
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "result",
            "modulus"
          ],
          "dest": "multiple",
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "n",
            "multiple"
          ],
          "dest": "remainder",
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "remainder"
          ],
          "op": "ret"
        }
      ],
      "name": "mod",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "two",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "args": [
            "n",
            "two"
          ],
          "dest": "bit",
          "funcs": [
            "mod"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "bit"
          ],
          "op": "ret"
        }
      ],
      "name": "lsb",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "n",
            "n"
          ],
          "dest": "result",
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "result"
          ],
          "op": "ret"
        }
      ],
      "name": "square",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "p",
          "type": "int"
        },
        {
          "name": "q",
          "type": "int"
        },
        {
          "name": "seed",
          "type": "int"
        },
        {
          "name": "sequence_length",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "p",
            "q"
          ],
          "dest": "modulus",
          "op": "mul",
          "type": "int"
        },
        {
          "dest": "start",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "sequence_length"
          ],
          "dest": "end",
          "op": "id",
          "type": "int"
        },
        {
          "label": "loop.branch"
        },
        {
          "args": [
            "start",
            "end"
          ],
          "dest": "cond",
          "op": "lt",
          "type": "bool"
        },
        {
          "args": [
            "cond"
          ],
          "labels": [
            "loop.body",
            "loop.end"
          ],
          "op": "br"
        },
        {
          "label": "loop.body"
        },
        {
          "args": [
            "seed"
          ],
          "dest": "squared",
          "funcs": [
            "square"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "squared",
            "modulus"
          ],
          "dest": "seed",
          "funcs": [
            "mod"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "seed"
          ],
          "dest": "least_significant_bit",
          "funcs": [
            "lsb"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "least_significant_bit"
          ],
          "op": "print"
        },
        {
          "dest": "one",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "start",
            "one"
          ],
          "dest": "start",
          "op": "add",
          "type": "int"
        },
        {
          "labels": [
            "loop.branch"
          ],
          "op": "jmp"
        },
        {
          "label": "loop.end"
        }
      ],
      "name": "main"
    }
  ]
}
