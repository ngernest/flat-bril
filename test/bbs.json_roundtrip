{
  "functions": [
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        },
        {
          "name": "modulus",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "n",
            "modulus"
          ],
          "dest": "result",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "result",
            "modulus"
          ],
          "dest": "multiple",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "n",
            "multiple"
          ],
          "dest": "remainder",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "remainder"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "mod",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "two",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "args": [
            "n",
            "two"
          ],
          "dest": "bit",
          "funcs": [
            "mod"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "bit"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "lsb",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "n",
            "n"
          ],
          "dest": "result",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "result"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "square",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "p",
          "type": "int"
        },
        {
          "name": "q",
          "type": "int"
        },
        {
          "name": "seed",
          "type": "int"
        },
        {
          "name": "sequence_length",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "p",
            "q"
          ],
          "dest": "modulus",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "dest": "start",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "sequence_length"
          ],
          "dest": "end",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "label": "loop.branch"
        },
        {
          "args": [
            "start",
            "end"
          ],
          "dest": "cond",
          "funcs": [],
          "labels": [],
          "op": "lt",
          "type": "bool"
        },
        {
          "args": [
            "cond"
          ],
          "funcs": [],
          "labels": [
            "loop.body",
            "loop.end"
          ],
          "op": "br"
        },
        {
          "label": "loop.body"
        },
        {
          "args": [
            "seed"
          ],
          "dest": "squared",
          "funcs": [
            "square"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "squared",
            "modulus"
          ],
          "dest": "seed",
          "funcs": [
            "mod"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "seed"
          ],
          "dest": "least_significant_bit",
          "funcs": [
            "lsb"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "least_significant_bit"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "dest": "one",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "start",
            "one"
          ],
          "dest": "start",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "loop.branch"
          ],
          "op": "jmp"
        },
        {
          "label": "loop.end"
        }
      ],
      "name": "main"
    }
  ]
}
