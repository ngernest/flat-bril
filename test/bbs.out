@mod
	op: div  		dest: result		type: int		args: ["n", "modulus"]	
	op: mul  		dest: multiple		type: int		args: ["result", "modulus"]	
	op: sub  		dest: remainder		type: int		args: ["n", "multiple"]	
	op: ret  		args: ["remainder"]	
@lsb
	op: const		dest: two		type: int		value: 2	
	op: call 		dest: bit		type: int		args: ["n", "two"]		funcs: ["mod"]
	op: ret  		args: ["bit"]	
@square
	op: mul  		dest: result		type: int		args: ["n", "n"]	
	op: ret  		args: ["result"]	
@main
	op: mul  		dest: modulus		type: int		args: ["p", "q"]	
	op: const		dest: start		type: int		value: 0	
	op: id   		dest: end		type: int		args: ["sequence_length"]	
.loop.branch
	op: lt   		dest: cond		type: bool		args: ["start", "end"]	
	op: br   		args: ["cond"]		labels: ["loop.body", "loop.end"]
.loop.body
	op: call 		dest: squared		type: int		args: ["seed"]		funcs: ["square"]
	op: call 		dest: seed		type: int		args: ["squared", "modulus"]		funcs: ["mod"]
	op: call 		dest: least_significant_bit		type: int		args: ["seed"]		funcs: ["lsb"]
	op: print		args: ["least_significant_bit"]	
	op: const		dest: one		type: int		value: 1	
	op: add  		dest: start		type: int		args: ["start", "one"]	
	op: jmp  		labels: ["loop.branch"]
.loop.end
